### PR TITLE
2.x: Rename create to unsafeCreate, add "safe" create methods.

### LINKED ISSUE #4286: 2.x: Add fromAsync Observable factory
but potentially with a different name based on what the resolution of #4255 is.

### PR DESCRIPTION
**Pull request by JakeWharton**
This is a proposal to solve #4255 but for 2.x and closes #4286 built around one goal: make the `create` methods actually usable by developers!

Currently the advice given to most developers using 1.x is avoid `create` unless you are absolutely 100% sure you understand backpressure and cancellation, and even then avoid it if possible. This method is a trap that most new developers fall into. _The method is named "create", after all, so it must be how I create these things!_ I want to make that statement not only correct, but advisable (at least when the still-convenient "from" methods are not suitable, such as when wrapping callback-based APIs).

Because `Observable`, `Single`, and `Completable` are backpressure-free, this amounts to just suppressing downstream notifications when disposed. This doesn't seem too controversial.

The most controversial part is that I have renamed `Flowable.fromAsync` to `Flowable.create` thus making it the canonical way for users to create `Flowable`s directly. I'm much less opinionated on this change, but it felt wrong to have an `Flowable.unsafeCreate` with no associated `create` and `fromAsync` is the "safest" version we have.

### CODE REVIEW

File: src/main/java/io/reactivex/internal/operators/completable/CompletableFromSource.java [hunk 1]
```diff
@@ -13,18 +13,62 @@
...
+import java.util.concurrent.atomic.AtomicBoolean;
 
+public final class CompletableFromSource extends CompletableFromUnsafeSource {
     public CompletableFromSource(CompletableSource source) {
-        this.source = source;
+        super(source);
+    }
+
+    @Override protected void subscribeActual(CompletableObserver observer) {
+        super.subscribeActual(new DisposeAwareCompletableObserver(observer));
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/completable/CompletableFromSource.java]: Why did you add this `DisposeAwareCompletableObserver`?

File: src/main/java/io/reactivex/internal/operators/completable/CompletableFromUnsafeSource.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+public class CompletableFromUnsafeSource extends Completable {
+
+    final CompletableSource source;
+
+    public CompletableFromUnsafeSource(CompletableSource source) {
+        this.source = source;
+    }
+    
+    @Override
+    protected void subscribeActual(CompletableObserver observer) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/completable/CompletableFromUnsafeSource.java:27]: I'd really would like these to remain abstract to force the subclasses to override it.

File: src/main/java/io/reactivex/internal/operators/observable/ObservableFromUnsafeSource.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+public class ObservableFromUnsafeSource<T> extends Observable<T> {
+    final ObservableSource<T> source;
+
+    public ObservableFromUnsafeSource(ObservableSource<T> source) {
+        this.source = source;
+    }
+    
+    @Override
+    protected void subscribeActual(Observer<? super T> observer) {
+        source.subscribe(observer);
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/observable/ObservableFromUnsafeSource.java:27]: I'd really would like these to remain abstract to force the subclasses to override it.

File: src/main/java/io/reactivex/internal/operators/single/SingleFromSource.java [hunk 4]
```diff
@@ -1,29 +1,72 @@
...
+
+    @Override protected void subscribeActual(SingleObserver<? super T> observer) {
+        super.subscribeActual(new DisposeAwareSingleObserver<T>(observer));
     }
-    
-    @Override
-    protected void subscribeActual(SingleObserver<? super T> observer) {
-        source.subscribe(observer);
+
+    static final class DisposeAwareSingleObserver<T>
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/single/SingleFromSource.java:36]: You could combine the single and completable variants into a single class.

File: src/main/java/io/reactivex/internal/operators/single/SingleFromUnsafeSource.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+public class SingleFromUnsafeSource<T> extends Single<T> {
+    final SingleSource<T> source;
+
+    public SingleFromUnsafeSource(SingleSource<T> source) {
+        this.source = source;
+    }
+    
+    @Override
+    protected void subscribeActual(SingleObserver<? super T> observer) {
+        source.subscribe(observer);
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/single/SingleFromUnsafeSource.java:27]: I'd really would like these to remain abstract to force the subclasses to override it.

File: src/main/java/io/reactivex/internal/operators/single/SingleFromSource.java (hunk 4 above)
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/single/SingleFromSource.java:36]: Sorry, forgot about the delegate type.

File: src/main/java/io/reactivex/internal/operators/observable/ObservableFromUnsafeSource.java (hunk 3 above)
Comment (JakeWharton) [inline src/main/java/io/reactivex/internal/operators/observable/ObservableFromUnsafeSource.java:27]: Yeah using inheritance here was an early crutch when prototyping. It's not needed in practice so I'll remove it.

File: src/main/java/io/reactivex/internal/operators/completable/CompletableFromSource.java (hunk 1 above)
Comment (JakeWharton) [inline src/main/java/io/reactivex/internal/operators/completable/CompletableFromSource.java]: It prevents notifications after disposal for implementations which might not be checking.

File: src/main/java/io/reactivex/internal/operators/observable/ObservableFromUnsafeSource.java (hunk 3 above)
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/observable/ObservableFromUnsafeSource.java:27]: Okay.

**Comment by JakeWharton** [comment]
I think it's also important to point out I replaced `Consumer<AsynEmitter<T>>` with `FlowableSource<T>` for parity with other "safe" `create` method parameter types. The unsafe variant still takes a raw `Publisher<T>`, however.

**Comment by artem-zinnatullin** [comment]
:heart::heart::heart:

**Comment by akarnokd** [comment]
:+1:
